### PR TITLE
Use npm workspaces for packages

### PR DESCRIPTION
**Pull request by t-hamano**
See this Slack comment:  https://wordpress.slack.com/archives/C02QB2JS7/p1727315839205439

> Does anyone know the best way to update existing dependencies?
> I'm referring to [this documentation](https://github.com/WordPress/gutenberg/blob/trunk/packages/README.md#updating-existing-dependencies), but the `lerna add {library} packages/{package-name}` command doesn't seem to work now.
> 
> ```
> lerna add change-case packages/a11y
> ERR! add The "add" command was removed by default in v7, and is no longer maintained.
> ERR! add Learn more about this change at https://lerna.js.org/docs/legacy-package-management
> ```
> 
> The npm command doesn't work either.
> 
> ```
> npm install change-case -w packages/components
> npm error No workspaces found:
> npm error   --workspace=packages/components
> ---
> npm install change-case -w components
> npm error No workspaces found:
> npm error   --workspace=components
> ```

## What?

This PR adds npm workspaces and updates the documentation to allow existing dependencies to be properly updated.

## Why?

lerna has a `lerna add` command to add dependencies to the monorepo. However, since npm native now has a similar feature, [this command was removed in version 7](https://lerna.js.org/docs/legacy-package-management). Therefore, the current documented method cannot properly update dependencies.

## How?

For npm publishing, we will continue to use lerna as before. This PR allows us to run npm native commands that are an alternative to `lerna add` by adding a `workspaces` to `package.json`.

lerna will refer to `workspaces` in `package.json` by default, but since the settings in `lerna.json` take precedence, this should not affect npm publishing ([Source](https://lerna.js.org/docs/api-reference/configuration#packages)).

## Testing Instructions

Please run a command like the following to make sure the dependencies are added properly:

```
npm install change-case -w packages/a11y
```

The script modules webpack build config should be identical. This can be verified by adding some logging and executing the file `node ./tools/webpack/script-modules.js`:

```diff
diff --git before/tools/webpack/script-modules.js after/tools/webpack/script-modules.js
index 3733b8954f..4d57d1e0dd 100644
--- before/tools/webpack/script-modules.js
+++ after/tools/webpack/script-modules.js
@@ -101,3 +101,5 @@ module.exports = {
 		aggregateTimeout: 500,
 	},
 };
+
+console.log( module.exports );
```

Closes #59933.

### CODE REVIEW

**Comment by t-hamano** [comment]
@gziolo @jsnajdr 

I'm trying to enable npm workspaces in the Gutenberg project so that dependencies are updated properly, but I'm running into two issues. If you have any approaches to resolve these, please let me know 🙇‍♂️

- When I run `npm install`, a lot of `"dev": true` are removed from `package-lock.json`. The difference is so large that it cannot be seen on GitHub, but you can check out this PR locally and run `git diff 93323fd1a3 HEAD -- package-lock.json` to see the difference.
- Probably due to the first effect, more library licenses are checked, causing the `other:check-licenses` command to fail. The ones that are considered incompatible with GPL2 are `Apache-2.0` and `CC-BY-3.0`.

**Comment by gziolo** [comment]
I’m curious if @kevin940726 run into the same issue when exploring namespaces in the past:
- https://github.com/WordPress/gutenberg/pull/48950

@sirreal or @desrosj, do you have some hints to share? We definitely need to have a way to distinguish production and development packages. Maybe React Native packages need to be explicitly marked as dev packages.

**Comment by sirreal** [comment]
Thanks for working on this, I think it needs to happen 🙂 

I need to investigate a bit more, but my first intuition is that maybe it's working correctly and the updates are as they should be 🤷 

For example, installing a new dev dependency that isn't currently present is correctly represented as a dev dep. Try this:

```sh
npm i -D oxlint -w @wordpress/a11y
```

**Comment by sirreal** [comment]
If we go in this direction, I wonder if the root package.json file should contain _much less_. An interesting question to consider is "what does the Gutenberg package.json represent?"

---

It's entirely possible this just works differently than lerna and it is correct under the npm workspaces model.

Here's a case: `@actions/core` had `dev: true` removed.

- It's a devDependency of the root. Maybe that's not necessary now? I think previously we hoisted all dev deps to the root because of lerna quirks, but I'm not sure.
- It's a dependency of 2 packages in the workspace: `@wordpress/project-management-automation` and `@wordpress/report-flaky-tests`.
- Both of those packages a dev dependencies of the root.

Does the package-lock represent the root package, or is it more like a union of everything in the workspace? It may be the latter, in that case since these are dependencies of a package in the workspace, they should be dev dependencies.

---

This is interesting:

```sh
npm i -D @actions/core
```

This changes the lockfile 🤦 it adds a license field in a few places where the package appears. I've never found npm to be very reliable and it continues today.

---

Here's an interesting diff, I compared the following command on trunk and the branch:

```sh
npm ls --package-lock-only --include=prod --parseable
```

<details>
<summary>diff of prod packages</summary>

```diff
diff --git before/trunk.txt after/branch.txt
index c8f325917e..f539055038 100644
--- before/trunk.txt
+++ after/branch.txt
@@ -69,6 +69,7 @@ node_modules/@wordpress/components
 node_modules/@wordpress/compose
 node_modules/@wordpress/core-commands
 node_modules/@wordpress/core-data
+node_modules/@wordpress/create-block-interactive-template
 node_modules/@wordpress/create-block-tutorial-template
 node_modules/@wordpress/create-block
 node_modules/@wordpress/customize-widgets
@@ -240,3 +241,176 @@ node_modules/webdriverio
 node_modules/webpack-bundle-analyzer
 node_modules/webpack
 node_modules/worker-farm
+node_modules/@babel/runtime
+packages/annotations/node_modules/uuid
+node_modules/deepmerge
+node_modules/gettext-parser
+node_modules/is-plain-object
+node_modules/@babel/plugin-transform-react-jsx
+node_modules/@babel/plugin-transform-runtime
+node_modules/@babel/preset-env
+node_modules/@babel/preset-typescript
+node_modules/core-js
+node_modules/clsx
+node_modules/@emotion/react
+node_modules/@emotion/styled
+node_modules/@react-spring/web
+node_modules/colord
+node_modules/diff
+node_modules/fast-deep-equal
+node_modules/memize
+node_modules/parsel-js
+node_modules/postcss-prefix-selector
+packages/block-editor/node_modules/postcss-urlrebase
+node_modules/react-autosize-textarea
+packages/block-editor/node_modules/react-easy-crop
+node_modules/remove-accents
+node_modules/fast-average-color
+packages/block-library/node_modules/uuid
+node_modules/pegjs
+node_modules/phpegjs
+node_modules/hpq
+packages/blocks/node_modules/react-is
+node_modules/showdown
+node_modules/simple-html-tokenizer
+packages/blocks/node_modules/uuid
+packages/commands/node_modules/cmdk
+packages/components/node_modules/@ariakit/react
+node_modules/@emotion/cache
+node_modules/@emotion/css
+node_modules/@emotion/serialize
+node_modules/@emotion/utils
+packages/components/node_modules/@floating-ui/react-dom
+packages/components/node_modules/@types/gradient-parser
+node_modules/@types/highlight-words-core
+node_modules/@use-gesture/react
+node_modules/date-fns
+node_modules/framer-motion
+node_modules/gradient-parser
+node_modules/highlight-words-core
+packages/components/node_modules/path-to-regexp
+node_modules/re-resizable
+node_modules/react-colorful
+packages/components/node_modules/uuid
+node_modules/@types/mousetrap
+packages/compose/node_modules/clipboard
+node_modules/mousetrap
+node_modules/use-memo-one
+packages/core-data/node_modules/uuid
+node_modules/check-node-version
+node_modules/mustache
+node_modules/npm-package-arg
+node_modules/write-pkg
+node_modules/is-promise
+node_modules/rememo
+packages/dataviews/node_modules/@ariakit/react
+node_modules/moment-timezone
+node_modules/moment
+node_modules/json2php
+node_modules/comment-parser
+node_modules/mdast-util-inject
+node_modules/optionator
+node_modules/remark-parse
+node_modules/remark
+node_modules/unified
+node_modules/form-data
+node_modules/get-port
+node_modules/lighthouse
+node_modules/mime
+packages/e2e-test-utils-playwright/node_modules/web-vitals
+node_modules/expect-puppeteer
+node_modules/jest-snapshot
+packages/e2e-tests/node_modules/uuid
+node_modules/copy-dir
+packages/env/node_modules/docker-compose
+node_modules/extract-zip
+node_modules/got
+node_modules/js-yaml
+node_modules/ora
+node_modules/terminal-link
+packages/env/node_modules/yargs
+node_modules/@babel/eslint-parser
+node_modules/@typescript-eslint/eslint-plugin
+node_modules/@typescript-eslint/parser
+node_modules/cosmiconfig
+node_modules/eslint-config-prettier
+node_modules/eslint-plugin-jsdoc
+node_modules/eslint-plugin-jsx-a11y
+node_modules/eslint-plugin-playwright
+node_modules/eslint-plugin-react-hooks
+node_modules/eslint-plugin-react
+node_modules/eslint
+node_modules/globals
+node_modules/requireindex
+node_modules/tannin
+packages/interactivity/node_modules/@preact/signals
+packages/interactivity/node_modules/preact
+node_modules/jest-matcher-utils
+node_modules/@axe-core/puppeteer
+node_modules/npm-package-json-lint
+node_modules/autoprefixer
+node_modules/requestidlecallback
+node_modules/@octokit/request-error
+packages/project-management-automation/node_modules/@octokit/webhooks
+node_modules/utility-types
+node_modules/@react-native-clipboard/clipboard
+node_modules/@react-native-community/blur
+node_modules/@react-native-community/slider
+node_modules/@react-native-masked-view/masked-view
+node_modules/@react-navigation/core
+node_modules/@react-navigation/native
+node_modules/@react-navigation/routers
+node_modules/@react-navigation/stack
+node_modules/jed
+node_modules/jsdom-jscore-rn
+node_modules/react-native-fast-image
+node_modules/react-native-gesture-handler
+node_modules/react-native-get-random-values
+node_modules/react-native-linear-gradient
+node_modules/react-native-modal
+node_modules/react-native-reanimated
+node_modules/react-native-safe-area-context
+node_modules/react-native-safe-area
+node_modules/react-native-sass-transformer
+node_modules/react-native-screens
+node_modules/react-native-svg
+packages/react-native-editor/node_modules/react-native-url-polyfill
+node_modules/react-native-video
+node_modules/react-native-webview
+node_modules/rungen
+packages/report-flaky-tests/node_modules/@actions/github
+node_modules/history
+node_modules/@svgr/webpack
+node_modules/adm-zip
+packages/scripts/node_modules/babel-loader
+packages/scripts/node_modules/chalk
+node_modules/clean-webpack-plugin
+node_modules/cross-spawn
+node_modules/cwd
+node_modules/dir-glob
+packages/scripts/node_modules/jest-dev-server
+node_modules/jest-environment-node
+node_modules/markdownlint-cli
+node_modules/merge-deep
+node_modules/mini-css-extract-plugin
+node_modules/minimist
+node_modules/npm-packlist
+node_modules/read-pkg-up
+node_modules/resolve-bin
+node_modules/rtlcss-webpack-plugin
+packages/scripts/node_modules/schema-utils
+node_modules/stylelint
+node_modules/url-loader
+node_modules/webpack-cli
+node_modules/webpack-dev-server
+node_modules/@stylistic/stylelint-plugin
+packages/stylelint-config/node_modules/stylelint-config-recommended-scss
+packages/stylelint-config/node_modules/stylelint-config-recommended
+node_modules/@types/simple-peer
+node_modules/import-locals
+node_modules/lib0
+node_modules/simple-peer
+node_modules/y-indexeddb
+node_modules/y-protocols
+node_modules/y-webrtc
+node_modules/yjs
```
</details>

**Comment by sirreal** [comment]
Here's a test, I added this package at `packages/test-package`:

```json
{
  "name": "test-package",
  "version": "1.0.0",
  "main": "index.js",
  "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1"
  },
  "author": "",
  "license": "ISC",
  "description": "",
  "devDependencies": {
    "ansi-align": "3.0.1"
  }
}
```

That's a new dependency. It's a dev dependency in the root. This is part of the updated package-lock:

```
		"ansi-align": {
			"version": "3.0.1",
			"resolved": "https://registry.npmjs.org/ansi-align/-/ansi-align-3.0.1.tgz",
			"integrity": "sha512-IOfwwBF5iczOjp/WeY4YxyjqAFMQoZufdQWDd19SEExbVLNXqvpzSJ/M7Za4/sCPmQ0+GRquoA7bGcINcxew6w==",
			"dev": true,
```

**Comment by sirreal** [comment]
Of course, the more interesting test is that a new prod dependency of any package in the workspace is a prod dependency in the root package json:

```diff
   "description": "",
-  "devDependencies": {
+  "dependencies": {
     "ansi-align": "3.0.1"
   }
 }
```

```diff
 		"ansi-align": {
 			"version": "3.0.1",
 			"resolved": "https://registry.npmjs.org/ansi-align/-/ansi-align-3.0.1.tgz",
 			"integrity": "sha512-IOfwwBF5iczOjp/WeY4YxyjqAFMQoZufdQWDd19SEExbVLNXqvpzSJ/M7Za4/sCPmQ0+GRquoA7bGcINcxew6w==",
-			"dev": true,
```

This does seem to indicate that the root package-lock functions as a union of workspace dependencies. This may be related to hoisting and other behaviors, but I think the behavior observed in this PR where many `"dev": true` are removed is correct.

**Comment by sirreal** [comment]
I created https://github.com/WordPress/gutenberg/pull/65708 (targeting this PR) with some ideas. I think we may want to move away from file refs now that we have workspaces. Lerna seems to handle it correctly in my testing.

**Comment by t-hamano** [comment]
Thanks for the detailed investigation!

If I understand correctly, the libraries that are causing a lot of `dev:true` to be removed, i.e. libraries that are defined as `devDependencies` in the root and as `dependencies` in the package, need to be moved to `dependencies` in the root as well or removed?

Please let me know if I can help you with anything.

**Comment by sirreal** [comment]
If we change to using workspaces —and I think we should— then it's really a rethinking of how the project is organized. There will be some learning along the way.

> libraries that are defined as devDependencies in the root and as dependencies in the package, need to be moved to dependencies in the root as well or removed?

I don't think this _needs_ to happen, but it seems to me that the root project effectively functions as an aggregate of all of the workspaces it contains. All of the workspaces are installed in the root `node_module` so I don't think any of them need to be declared directly as dependencies.

I think it would be good to wait until this release is settled (to prevent disruption at a critical time) and then explore some of these changes.

**Comment by gziolo** [comment]
> I don't think this needs to happen, but it seems to me that the root project effectively functions as an aggregate of all of the workspaces it contains

That's a good point. It seems like a correct assumption that all WP packages would be now in the workspace so we no longer need to reference them from the root `package.json` file. That said, in the individual packages, we still need to use `file:path/to/package` references so all the packages are consumed from the workspace. At least, this is my understanding how these packages should be linked between themselves.

The development packages should continue to be maintained from the root of the project as explained https://github.com/WordPress/gutenberg/tree/trunk/packages#development-dependencies. The biggest challenge moving forward is going to be how to configure `wp-scripts check-licenses` so it correctly recognizes which packages are used on production vs these that are for development purposes. So far, that decision making process was happening through `package.json` file by configuring `dependencies` vs `devDependencies`. The method used behind the scenes is `npm ls` which probably depends on the `node_modules` folder or the lock file. I don't know if that has changed much with workspaces, or the workspaces just surface some inherent issues with the lock file that existed before.

**Comment by sirreal** [comment]
> That said, in the individual packages, we still need to use file:path/to/package references so all the packages are consumed from the workspace. At least, this is my understanding how these packages should be linked between themselves.

I was curious about this and it was something I explored in https://github.com/WordPress/gutenberg/pull/65708. Lerna's docs use `*` I believe, and when I tried some lerna commands locally, it worked as expected (lerna replaced the `*` versions with the local versions). Inside `node_modules`, the workspaced packages are linked, so I don't think we need to use file refs anymore.

We should definitely do more testing there.

**Comment by gziolo** [comment]
> Lerna's docs use * I believe, and when I tried some lerna commands locally, it worked as expected (lerna replaced the * versions with the local versions). 

That would work assuming that `*` gets replaced by Lerna during publishing with the correct version ranges.

**Comment by gziolo** [comment]
@ciampo, @jsnajdr and @tyxla, I see you working this week on some updates for dependencies to npm packages. I was wondering if you have enough insights to help with this one. It potentially could simplify the process of updating dependencies on packages.

**Comment by ciampo** [comment]
I personally don't have many insights, apart from the fact that currently updating dependencies is not a simple task.

The only reliable way, for me, is to remove the dependencies, run `npm install`, re-add the dependencies (with the new version), and re-run `npm install`.

So any changes that can improve the process would be great!

**Comment by jsnajdr** [comment]
I tried to figure out why all the `dev: true` fields disappear and I can more or less confirm what @sirreal says. It's the tooling packages like `@wordpress/eslint-plugin` or `@wordpress/e2e-test-utils`. Until now they were dev dependencies of the root `package.json`, and their transitive non-dev dependencies were marked as `dev: true` because they were "strictly dev".

But now these packages are included by virtue of being in `workspaces`, they are not flagged as `dev` any more, so their dependences are not `dev: true` either.

So, the disappearance of `dev: true` is legit and harmless. We should however remove all the "local" dependencies from the root `package.json`, they are not needed any more, the job is now done by the `workspaces` field.

**Comment by jsnajdr** [comment]
Few other things we should probably do when enabling npm workspaces for the repo:

Migrate calls of `lerna run build` or `lerna run postinstall` to the equivalent `npm run build --workspaces --if-present`. That should do the same job natively: find subpackages that have the `build` script and run them.

After this, we'll be using Lerna only for the publishing workflow: `lerna version`, `lerna updated`, `lerna publish`. Is Lerna still the best tool for this or is there a replacement? Doesn't look like npm supports these things natively.

Remove individual `.npmrc` files from the `packages/*` directories. They specify `package-lock: false`, for some reason, and workspace-enabled npm explicitly checks for their presence and warns that they are going to be ignored:
```
 npm warn ignoring workspace config at packages/block-serialization-spec-parser/.npmrc
```

**Comment by gziolo** [comment]
> After this, we'll be using Lerna only for the publishing workflow: lerna version, lerna updated, lerna publish. Is Lerna still the best tool for this or is there a replacement? Doesn't look like npm supports these things natively.

This is how it should be. Lerna is only useful for npm publishing these days. I don't know if still the best tool for the job, but the amount of work necessary to migrate to any other tool is probably not worth it. I personally won't have availability to help with such migration in the near future.

> Remove individual .npmrc files from the packages/* directories. They specify package-lock: false, for some reason, and workspace-enabled npm explicitly checks for their presence and warns that they are going to be ignored:

This flag currently prevents creating lock files in each individual package which is no longer necessary with workspaces as it uses a single lock file. You simply don't need these lock files with monorepo so they would only create noise.

**Comment by gziolo** [comment]
> So, the disappearance of dev: true is legit and harmless. We should however remove all the "local" dependencies from the root package.json, they are not needed any more, the job is now done by the workspaces field.

While it's perfectly fine. It means we will have to fix the script that checks licenses. I think the simplest way will be ignoring packages that are used only for dev purposes, similar to how it happens today with some packages:

https://github.com/WordPress/gutenberg/blob/ec5acbdbe0ccffd734de3f2aa76b42bb974b79b7/package.json#L308

Example for checking production:

```bash
wp-scripts check-licenses --prod --gpl2 --ignore=@wordpress/scripts,@wordpress/babel-preset-default,@wordpress/...
```

It isn't entirely clear how the same can be achieved for dev packages that will be marked now as production packages, maybe:

```bash
wp-scripts check-licenses --prod --ignore=@wordpress/block-edtior,@wordpress/blocks,@wordpress/...
```

In effect, we might need three parallel scripts moving forward. It's also possible that `--dev` covers all packages, no matter they define `dev:true` or `dev:false`.

**Comment by jsnajdr** [comment]
> It means we will have to fix the script that checks licenses.

We want to check licenses only for packages that are going to be included in the Gutenberg plugin, right? Not for other ones that are merely published to NPM.

This is very tightly coupled to how the packages are listed and built in `tools/webpack/packages.js`. Until now the source of truth for the packages built for the plugin was the `dependencies` list in the root `package.json`. The webpack config reads that list and constructs the list of entrypoints to build. Also the `check-licenses` script uses the `dependencies` list, indirectly, as it is represented in the `npm ls` output.

Now when the top-level dependencies are going away, replaced by workplaces, we need to find a new place for this source of truth -- where the production `@wordpress/*` packages are listed.

**Comment by sirreal** [comment]
With Gutenberg 19.4 recently released, It should be an ideal time to work on this. 

https://github.com/WordPress/gutenberg/pull/65708 is likely a good start.

**Comment by gziolo** [comment]
> We want to check licenses only for packages that are going to be included in the Gutenberg plugin, right? Not for other ones that are merely published to NPM.

Unfortunately, it is not sufficient to validate all dependencies bundled on code shipped on production. That part needs to be GPL compatible. Everything else used in this repository also needs to have a proper license, but it's less strict, and a wider set of licenses is allowed.

**Comment by sirreal** [comment]
There's some work in https://github.com/WordPress/gutenberg/pull/59933 to switch to npm workspaces.

**Comment by sirreal** [comment]
The Gutenberg plugin was released 5 days ago, package releases should happen this Wednesday. After the next package release would be a great time to switch to workspaces so that there's a longer cycle for testing.

I'll do some work on this PR this week trying to get ready.

**Comment by sirreal** [comment]
I'll continue using this PR to maintain context, but I'll rebase and push this forward incorporating ideas from #65708. Before making major changes the commit was d99f87c00c951ef72900fb09a0837568ead8b63a.

File: tools/webpack/packages.js
Comment (sirreal) [inline tools/webpack/packages.js]: Similar to script modules, maybe a field in a package's package.json can indicate it should be included.

Probably not just this list (copied from the devDependencies on trunk)

**Comment by sirreal** [comment]
I think the way to handle packages that are included in WordPress Core (for licensing and for bundling) is by adding a field to their package JSON.

Script modules have that field now, `wpScriptModuleExports`:

https://github.com/WordPress/gutenberg/blob/1e8f5e01d7d2f43cd16f00aef03f3d16f543ae2f/packages/interactivity/package.json#L29-L32

I'll add `"wpScript": true` to packages that are included in Gutenberg/Core as scripts. This should solve some of the problems with packaging scripts and with checking licenses.

The check license script will need to change. I plan to separate the license checking part from the package reading part. Third partied will be able to use the license checking flow. Gutenberg will use the same license checking logic, but will get packages from a query starting like: `npm query '.workspace:attr([isWpScript]), .workspace:attr([wpScriptModuleExports])'`

File: package-lock.json
Comment (gziolo) [inline package-lock.json:51599]: Finally, I see a lot of packages hoisted after changes are applied with this PR, but there are still some places where npm packages are installed inside the local package. I hope we can remove all these instances over time.

Comment (sirreal) [reply package-lock.json:51599]: Some duplication is likely inevitable because it's very difficult to get all transitive dependencies onto the same version of a package all the time. When incompatible versions manifest, I think that's always going to cause deps to be installed under packages.

Hopefully this will get us to a more sane place of dependency management. This is in a place where I can try one of the dependency upgrades that was constantly stuck and see if it works better (e.g. #61628).

File: bin/check-licenses.mjs
Comment (sirreal) [inline bin/check-licenses.mjs]: This list is moved from the ignored list that was in package.json checking prod dependencies. What were prod dependencies are now the `wpScript` workspace packages.

**Comment by sirreal** [comment]
I think this is working correctly now and can be reviewed. I'm sure there will be changes to make to the workspaces setup, but all of the main pieces are in this PR.

File: package.json
Comment (gziolo) [inline package.json:24]: I’m wondering whether there are still any scripts that used this list of dependencies to do some processing. I need to think how to verify it.

**Review by gziolo — COMMENTED** [review]
Overall, it’s looking great. I still would like to scan the codebase looking for similar patterns as in the webpack configs where dependencies were sourced from the `package.json` file.

File: bin/check-licenses.mjs
Comment (gziolo) [reply bin/check-licenses.mjs]: Is is still needed when RN dependencies are no longer listed as production?

Comment (sirreal) [reply bin/check-licenses.mjs]: 👍 I updated the list to remove all entries that do not error.

File: bin/check-licenses.mjs
Comment (gziolo) [inline bin/check-licenses.mjs:13]: I see it’s licensed under [Apache-2.0 license](https://github.com/ampproject/remapping/blob/main/LICENSE) so it would be handy to add now a comment explaining why it can be ignored. I hope it’s only listed as a dependency of a dependency but never used.

File: package.json
Comment (gziolo) [reply package.json:24]: I couldn't identify any occurences. It looks like everything was refactored successfully.

File: bin/check-licenses.mjs
Comment (gziolo) [reply bin/check-licenses.mjs:13]: So this is a dependency of `@babel/core`. The only production package that depends on `@babel/core` that I could locate is `@emotion/css`. It's used in a single place:

https://github.com/WordPress/gutenberg/blob/b8cd03d2e2713ffe648923712d2599b25808ad67/packages/components/src/utils/hooks/use-cx.ts#L7-L10

It doesn't seem to use Babel at all:

> `cx` is emotion's version of the popular [`classnames` library](https://github.com/JedWatson/classnames).

The latest version of `@emotion/css` doesn't define `@babel/core` as a peer dependency anymore:

https://github.com/emotion-js/emotion/blob/main/packages/css/package.json

It has changed with the following PR: https://github.com/emotion-js/emotion/pull/2985. The package was at `11.10.5` at the time of that change. The version used in the Gutenberg repository is `11.7.1`. It could be that bumping these dependencies would make this ignore obsolete.

Comment (sirreal) [reply bin/check-licenses.mjs:13]: That would be nice, but I'd prefer to leave upgrades for after this PR has landed. It's already quite a big and potentially disruptive PR.

Comment (gziolo) [reply bin/check-licenses.mjs:13]: I wanted to document everything as I was digging into the details of prior decisions. There is no need for any follow-up in this PR.

**Comment by gziolo** [comment]
I did some testing and there are some nice possibilities introduced.

1. Installs the package with the same version as in the lock file:

```
npm install change-case -w packages/a11y
```

2. Enforce installing the latest version:

```
npm install change-case@latest -w packages/a11y
```

3. Install in all packages:

```
npm install change-case@latest --workspaces
```

Aside. I'm not sure whether workspaces solve the problem when a given dependency needs to be updated to a newer version in several packages and hoisted. In example (2), the latest version lands in the `packages/a11y/node_modules` subfolder.

**Comment by gziolo** [comment]
Interesting discovery. There are at least two packages that we bundle with webpack in Gutenberg but they are never consumed. It isn't specific to this PR though:
- `@wordpress/block-serialization-spec-parser`
- `@wordpress/react-i18n`

Both are annotated with `wpScripts`.

**Review by gziolo — APPROVED** [review]
It works great in my testing. Let's land it in the current shape.

I'm already working on a follow-up to enforce a consistent order of new properties in the `package.json` file. I will also take a closer look at the existing documentation to see how we can improve it and put some details about:
- `wpScripts`
- `wpScriptModuleExports`

**Comment by t-hamano** [comment]
I really appreciate you working on this PR. All I can do is test it on a Windows OS, the following two commands complete successfully:

- `npm run build`
- `npm run other:check-licenses`
- `npm install change-case -w packages/a11y`
- `npm install change-case@latest -w packages/a11y`
- `npm install change-case@latest --workspaces`

**Comment by gziolo** [comment]
First follow-up PR is ready:
- https://github.com/WordPress/gutenberg/pull/66239

**Comment by gziolo** [comment]
All `.npmrc` files were removed, so we should also update the docs that suggested creating this file for every new package. See [c676375](https://github.com/WordPress/gutenberg/pull/65681/commits/c676375db6ae389215cb187c784648a1d8302d65).

**Review by sirreal — APPROVED** [review]
I think this is ready to go.

I would avoid landing it until after the weekend in case some issues come up.

I'd like to include the props from the workspaces PR this superseded: https://github.com/WordPress/gutenberg/pull/59933#issuecomment-2002321944

**Comment by gziolo** [comment]
I played a bit more with workspaces and the result it the following documentation edit: https://github.com/WordPress/gutenberg/pull/65681/commits/d10c7d68d23f42fb2c607f42375ff1a76c226717.

**Pull request merged**
